add accesToken to client + add limit to getLastPlayed

melosh101 opened the pull request:
add caching of the accesToken to the client
add limit to the getLastPlayed function to allow the user to pick the amount of tracks to receive. max being 50 and minimum being 1.
add test to test the limiting feature on getlastPlayed
fixes #10
fixes #7

rocktimsaikia commented on index.ts:
```diff
@@ -55,21 +56,22 @@ export class SpotifyClient {
       })
     });
     const responseData = (await response.json()) as AccessToken;
-    return responseData?.access_token;
```
Let's not modify this right here. The function should only do one thing. In this case, It should only give the accessToken so that we don't have to worry about its returned data having other properties. Makes debugging easy :)

rocktimsaikia commented on index.ts:
```diff
@@ -55,21 +56,22 @@ export class SpotifyClient {
-      const headers = { Authorization: `Bearer ${accessToken}` };
+      if(this.accessToken === null) 
+        this.accessToken = await this._genAccesToken();
+      const headers = { Authorization: `Bearer ${this.accessToken.access_token}` };
```
this will also change to `this.accessToken` after the above-commented change.

rocktimsaikia commented on index.ts:
```diff
@@ -80,14 +82,23 @@ export class SpotifyClient {
+      if(this.accessToken === null) 
+        this.accessToken = await this._genAccesToken();
+      
+      if(limit > 50 || limit < 1)
+        throw new Error('Limit must be between 1 and 50');
```
Why this specific limit? Just curious

melosh101 replied:
this is the limit specified in the Spotify docs
https://developer.spotify.com/documentation/web-api/reference/#/operations/get-recently-played

rocktimsaikia commented on index.ts:
```diff
@@ -80,14 +82,23 @@ export class SpotifyClient {
+      if(limit > 50 || limit < 1)
+        throw new Error('Limit must be between 1 and 50');
+      
+      const headers = { Authorization: `Bearer ${this.accessToken.access_token}` };
```
this will also change to `this.accessToken`

rocktimsaikia commented on index.ts:
```diff
@@ -80,14 +82,23 @@ export class SpotifyClient {
       // TODO: add types once spotify-types is updated
       const responseData = (await response.json()) as any;
-      const lastPlayedTrack = filterResponse(responseData?.items[0]?.track);
+      const lastPlayedTrack = responseData.items.map((item: { track: Track; }) => {
+        const track = filterResponse(item.track);
+        return track;	
+      })
```
We can do this I guess. Less verbose.
```ts
const lastPlayedTrack = responseData.items.map(({ track }: { track: Track }) => {
      return filterResponse(track);
 });
```

PS: We will remove typing on this map in future once this [issue](https://github.com/spotify-api/spotify-types/issues/4) is resolved in `spoity-types` lib

melosh101 replied:
i agree. looks more clean, dont know why i didnt do that from the start

rocktimsaikia commented on index.ts:
```diff
@@ -55,21 +56,22 @@ export class SpotifyClient {
         isPlaying = false;
         return fallbackToLastPlayed
-          ? { isPlaying, ...(await this.getLastPlayed()) }
+          ? { isPlaying, ...(await this.getLastPlayed())[0] as ResponseTrack }
```
the type here is already inferred with the return type of `getLastPlayed`. I don't need to think we need to explicitly add typing here

melosh101 replied:
you're right. this is not necessary and should be removed

rocktimsaikia commented on index.ts:
```diff
@@ -80,14 +82,23 @@ export class SpotifyClient {
   };
 
-  getLastPlayed = async () => {
+  getLastPlayed = async (limit: number = 1): Promise<ResponseTrack[]> => {
```
After this change, this will return an Array instead of an object. This is a breaking change and needs to be updated in the readme example too.

rocktimsaikia reviewed (changes requested):
Good job opening this PR this soon. Left some comments that need to be addressed before merging!

rocktimsaikia commented:
Another point to note is that your commit message should be a one-liner most of the time explaining what it's doing for example:
```md
cache accesToken, add limit option to getLastPlayed
```
 and when you actually create the PR, you can explain the feature details and add `fixes #<id>` statements in the PR description.

melosh101 commented:
i have added the missing 401 if clauses and added a use case of the limit option on getLastPlayed function.
addressed the changes to getLastPlayed returning a array instead of an object

rocktimsaikia commented on readme.md:
```diff
@@ -42,11 +42,30 @@ const currentlyPlayingTrack = await spotify.getCurrentlyPlaying();
+ ]
+*/
+
+// If you wish to get a list of the recently played songs, pass the limit option, this can be an integer between 1 and 50
```
```suggestion
// To get a specific number of the recently played songs, just pass it to the method (1 < n < 50), default is 1
```
